File: app/src/main/java/com/divercity/app/features/profile/selectgroups/SelectGroupFragment.java
package com.divercity.app.features.profile.selectgroups;


import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProviders;
import android.arch.paging.PagedList;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.divercity.app.R;
import com.divercity.app.core.base.BaseFragment;
import com.divercity.app.core.ui.NetworkState;
import com.divercity.app.core.ui.RetryCallback;
import com.divercity.app.core.utils.Preconditions;
import com.divercity.app.data.Status;
import com.divercity.app.data.entity.group.GroupResponse;
import com.divercity.app.databinding.FragmentSelectGroupBinding;
import com.divercity.app.features.profile.selectgroups.adapter.GroupsAdapter;
import com.divercity.app.features.profile.selectgroups.adapter.GroupsViewHolder;

import javax.inject.Inject;

import timber.log.Timber;

public class SelectGroupFragment extends BaseFragment implements RetryCallback {

    private static final int PROGRESS = 95;

    SelectGroupViewModel viewModel;
    FragmentSelectGroupBinding binding;
    EditText edTxtSearch;

    int currentPosition;

    @Inject
    public GroupsAdapter groupsAdapter;
    private boolean isRefreshing = false;

    public static SelectGroupFragment newInstance() {
        return new SelectGroupFragment();
    }

    public SelectGroupFragment() {
        // Required empty public constructor
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_select_group;
    }

    @Override
    public ViewModel getViewModel() {
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(SelectGroupViewModel.class);
        return viewModel;
    }

    @Override
    public int getBindingVariable() {
        return -1;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = (FragmentSelectGroupBinding) getViewDataBinding();
        viewModel.fetchCompanies(this, null);
        groupsAdapter.setRetryCallback(this);
        groupsAdapter.setListener(listener);
        binding.listGroup.setAdapter(groupsAdapter);
        subscribeToJoinLiveData();
        subscribeToLiveDataPaginated();
        setupHeader();

        binding.btnContinue.setOnClickListener(view1 -> {
            navigator.navigateToNextProfile(getActivity(), viewModel.getAccountType());
        });
    }

    private void setupHeader(){
        edTxtSearch = binding.includeSearch.edtxtSearch;
        edTxtSearch.setOnKeyListener((v, keyCode, event) -> {
            // If the event is a key-down event on the "enter" button
            if ((event.getAction() == KeyEvent.ACTION_DOWN) &&
                    (keyCode == KeyEvent.KEYCODE_ENTER)) {
                String toSearch = edTxtSearch.getText().toString();
                if (toSearch.equals(""))
                    toSearch = null;
                viewModel.fetchCompanies(SelectGroupFragment.this,
                        toSearch);
                subscribeToLiveDataPaginated();
                return true;
            }
            return false;
        });

        binding.includeHeader.progressBar.setMax(100);
        binding.includeHeader.progressBar.setProgress(0);
        binding.includeHeader.progressBar.setProgressWithAnim(PROGRESS);
        binding.includeHeader.txtProgress.setText(PROGRESS + "%");
        binding.includeHeader.txtTitle.setText(R.string.select_group);

        binding.includeHeader.btnClose.setOnClickListener(view1 -> {

        });

        binding.includeHeader.btnSkip.setOnClickListener(view1 -> {

        });
    }

    private void subscribeToLiveDataPaginated() {
        viewModel.companyList.observe(this, new Observer<PagedList<GroupResponse>>() {
            @Override
            public void onChanged(@Nullable PagedList<GroupResponse> interestItems) {
                Timber.e("questionList setting list: " + interestItems.size());
                groupsAdapter.submitList(interestItems);
                if (groupsAdapter.getCurrentList() != null)
                    Timber.e("questionList CURRENT LIST: " + groupsAdapter.getCurrentList().size());
            }
        });

        viewModel.getNetworkState().observe(this, new Observer<NetworkState>() {
            @Override
            public void onChanged(@Nullable NetworkState networkState) {
                if (groupsAdapter.getCurrentList() != null)
                    Timber.e("CURRENT LIST: " + groupsAdapter.getCurrentList().size());

                if (!isRefreshing || networkState.getStatus() == Status.ERROR || networkState.getStatus() == Status.SUCCESS)
                    groupsAdapter.setNetworkState(networkState);
            }
        });

        viewModel.getRefreshState().observe(this, new Observer<NetworkState>() {
            @Override
            public void onChanged(@Nullable NetworkState networkState) {
                if (groupsAdapter.getCurrentList() != null)
                    Timber.e("CURRENT LIST: " + groupsAdapter.getCurrentList().size() + " y " + networkState.getStatus().name());
                if (networkState != null) {
                    Timber.e("STATUS: " + networkState.getStatus().name());
                    if (groupsAdapter.getCurrentList() != null) {
                        if (networkState.getStatus()
                                == NetworkState.LOADED.getStatus() || networkState.getStatus()
                                == Status.ERROR)
                            isRefreshing = false;
                        Timber.e("STATUS size: " + groupsAdapter.getCurrentList().size());

                        boolean d = networkState.getStatus()
                                == NetworkState.LOADING.getStatus();
                        Timber.e("STATUS: " + d);
                    }

                    if (networkState.getStatus() == Status.SUCCESS &&
                            groupsAdapter.getCurrentList().size() == 0) {
                        binding.txtNoResults.setVisibility(View.VISIBLE);
                    } else {
                        binding.txtNoResults.setVisibility(View.GONE);
                    }
                }
            }
        });
    }

    private void subscribeToJoinLiveData(){
        viewModel.joinGroupData.observe(this, response -> {
            Preconditions.checkNotNull(response);
            binding.includeLoading.setResource(response);
            switch (response.status){
                case ERROR:
                    Toast.makeText(getActivity(),response.message,Toast.LENGTH_SHORT).show();
                    break;
                case SUCCESS:
                    groupsAdapter.getCurrentList().get(currentPosition).getAttributes().setIsFollowedByCurrent(true);
                    groupsAdapter.notifyItemChanged(currentPosition);
                    break;
            }
        });
    }

    @Override
    public void retry() {
        viewModel.retry();
    }

    private GroupsViewHolder.Listener listener = (position, group) ->{
        currentPosition = position;
        viewModel.joinGroup(group);
    };
}

File: app/src/main/java/com/divercity/app/features/onboarding/selectgroups/SelectGroupFragment.kt
package com.divercity.app.features.onboarding.selectgroups

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.KeyEvent
import android.view.View
import android.widget.Toast
import com.divercity.app.R
import com.divercity.app.core.base.BaseFragment
import com.divercity.app.core.ui.RetryCallback
import com.divercity.app.core.utils.OnboardingProgression
import com.divercity.app.data.Status
import com.divercity.app.features.onboarding.selectgroups.adapter.GroupsAdapter
import com.divercity.app.features.onboarding.selectgroups.adapter.GroupsViewHolder
import kotlinx.android.synthetic.main.fragment_onboarding_header_search_list.*
import kotlinx.android.synthetic.main.view_header_profile.*
import kotlinx.android.synthetic.main.view_search.view.*
import javax.inject.Inject

/**
 * Created by lucas on 25/10/2018.
 */


class SelectGroupFragment : BaseFragment(), RetryCallback {

    lateinit var viewModel: SelectGroupViewModel

    @Inject
    lateinit var adapter: GroupsAdapter

    private var positionJoinClicked: Int = 0
    private var countJoin = 0
    private var currentProgress: Int = 0

    companion object {
        private const val PARAM_PROGRESS = "paramProgress"

        fun newInstance(progress: Int): SelectGroupFragment {
            val fragment = SelectGroupFragment()
            val arguments = Bundle()
            arguments.putInt(PARAM_PROGRESS, progress)
            fragment.arguments = arguments
            return fragment
        }
    }

    override fun layoutId(): Int = R.layout.fragment_onboarding_header_search_list

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(this, viewModelFactory)[SelectGroupViewModel::class.java]
        currentProgress = arguments?.getInt(PARAM_PROGRESS) ?: 0
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.fetchCompanies(this, null)
        adapter.setRetryCallback(this)
        adapter.setListener(listener)
        list.adapter = adapter
        subscribeToPaginatedLiveData()
        subscribeToJoinLiveData()
        setupHeader()
    }

    private fun setupHeader() {
        include_search.edtxt_search.setOnKeyListener { _, keyCode, keyEvent ->
            if (keyEvent.getAction() == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_ENTER) {

                var toSearch: String? = include_search.edtxt_search.getText().toString()

                if (toSearch == "")
                    toSearch = null

                viewModel.fetchCompanies(this@SelectGroupFragment, toSearch)
                subscribeToPaginatedLiveData()
                true
            } else
                false
        }

        include_header.apply {

            progress_bar.apply {
                max = 100
                progress = 0
                setProgressWithAnim(currentProgress)
            }
            txt_title.setText(R.string.select_group)

            txt_progress.text = currentProgress.toString().plus("%")

            btn_close.setOnClickListener {
                navigator.navigateToHomeActivity(activity!!)
            }

            btn_skip.setOnClickListener {
                navigator.navigateToHomeActivity(activity!!)
            }
        }
    }

    private fun subscribeToJoinLiveData() {
        viewModel.joinGroupResponse.observe(this, Observer { school ->
            when (school?.status) {
                Status.LOADING -> showProgress()

                Status.ERROR -> {
                    hideProgress()
                    Toast.makeText(activity, school.message, Toast.LENGTH_SHORT).show()
                }
                Status.SUCCESS -> {
                    hideProgress()
                    // Updating join btn state
                    adapter.currentList?.get(positionJoinClicked)?.attributes?.isIsFollowedByCurrent = true
                    adapter.notifyItemChanged(positionJoinClicked)

                    if (++countJoin == 3) {
                        btn_continue.background = ContextCompat.getDrawable(activity!!, R.drawable.shape_backgrd_round_blue2)
                        btn_continue.setOnClickListener {
                            val nextProgress = OnboardingProgression.getNextNavigationProgressOnboarding(
                                    activity!!, viewModel.accountType, currentProgress
                            )
                            navigator.navigateToHomeActivity(activity!!, nextProgress >= 100)
                        }
                    }
                }
            }
        })
    }

    private fun subscribeToPaginatedLiveData() {
        viewModel.pagedGroupList.observe(this, Observer {
            adapter.submitList(it)
        })

        viewModel.networkState.observe(this, Observer {
            adapter.setNetworkState(it)
        })

        viewModel.refreshState.observe(this, Observer { networkState ->
            networkState?.let {
                adapter.currentList?.let {
                    if (networkState.status == Status.SUCCESS && it.size == 0)
                        txt_no_results.visibility = View.VISIBLE
                    else
                        txt_no_results.visibility = View.GONE
                }
            }
        })
    }

    override fun retry() {
        viewModel.retry()
    }

    private val listener: GroupsViewHolder.Listener = GroupsViewHolder.Listener { position, group ->
        positionJoinClicked = position
        viewModel.joinGroup(group)
    }
}